.github/workflows/publish.yaml

```diff
@@ -13,11 +13,13 @@ env:
 concurrency:
   group: "publish"
   cancel-in-progress: false
-
 jobs:
   publish-python:
     runs-on: ubuntu-22.04  # convco needs GLIBC_2.32 which is not in 20.04
     environment: publish
+    # IMPORTANT: trusted publishing requires permission: id-token: write! Without it, the trusted publishing will not work!
+    # If the permission is the only one that is set, all other permissions are automatically set to none, which may cause permissions issues within the job.
```

Comment:
@jzizka91 @mbelak-dtml  I do not understand the last sentence (If the permission is the only one that si set...). Does it mean that we have to set `permissions: write-all` instead of `permissions: write`?

Reply:
@lukany 

By default permissions in all action are permissive meaning almost all permission scopes in workflows have read/write by default. Except for `id-token` and `metadata` scope which has `none` permissions. Usually you don't need to worry about permissions in workflows unless you need to modify permission of some scope. For example like in this workflow or specifically that `publish-python` job. To make publish with trusted-publishing method possible, permission scope `id-token` (which is by default none) needs to be set to `write`, otherwise it would not work. 

Here is the catch. If you specify any permission e.g. `id-token: write`, all other permission scopes that are not specified are set to none. This is by design. You can read about [here](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs) in third paragraph. 

For example, if you would specify only: `permissions: id-token: write` than actions like `checkout/v4` would simply not work because permission scope `contents` that requires `read` to make the action work would be `none`

You can read about permission scopes [here](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs#overview) under the review section.

Now, the reason why it was set to `permissions: write-all` is because I wasn't completely sure exactly what permission scopes (besides `id-token`) are needed to maintain the full functionality of the job. After little experiencing I found out what permission scopes are needed for some operations. Like the `contents` scope that allows actions like  `checkout/v4` to work within a job.

After I found this out I suggested to MichalB to modify this permissions to following:

```yml
permissions:
  id-token: write
  contents: read
  packages: read 
```

`packages: read` is optional. It would be required only if the job was using a container image (e.g. `container: ghcr.io/datamole-ai/core-tools-pythonruntime:11.1.0`), but since this job is not using it can be omitted.

The commet section for the permissions I used can be removed or modified to your liking.  I'm sorry for not making the comment more clear :)

Reply:
Ah, I understand now and it makes sense. Thanks! Appreciate you took the time to fully understand the problem and also explain it here!

I think my misunderstanding was mostly caused my lack of knowledge about the permissions. The comment is good IMO.